extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -1696,111 +1691,128 @@ private KafkaSupervisorReport generateReport(boolean includeOffsets)
-            lagComputingConsumer.seekToEnd(ImmutableList.of(topicPartition));
-            if (offsetsResponse.get(topicPartition.partition()) != null) {
-              diff = lagComputingConsumer.position(topicPartition) - offsetsResponse.get(topicPartition.partition());
-              lastCurrentOffsets.put(topicPartition.partition(), offsetsResponse.get(topicPartition.partition()));
-            } else if (lastCurrentOffsets.get(topicPartition.partition()) != null) {
-              diff = lagComputingConsumer.position(topicPartition) - lastCurrentOffsets.get(topicPartition.partition());
-            } else {
-              throw new ISE("Could not find latest consumed offset for partition [%d]", topicPartition.partition());
-            }
-            lag += diff;
-            log.debug(
-                "Topic - [%s] Partition - [%d] : Partition lag [%,d], Total lag so far [%,d]",
-                topicPartition.topic(),
-                topicPartition.partition(),
-                diff,
-                lag
-            );
-          }
-          emitter.emit(
-              ServiceMetricEvent.builder().setDimension("dataSource", dataSource).build("ingest/kafka/lag", lag)
-          );
-        }
-        catch (InterruptedException e) {
-          // do nothing, probably we are shutting down
+    return taskGroups
+        .values()
+        .stream()
+        .flatMap(taskGroup -> taskGroup.tasks.entrySet().stream())
+        .flatMap(taskData -> taskData.getValue().currentOffsets.entrySet().stream())
+        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, Long::max));
+  }
+
+  private Map<Integer, Long> getLagPerPartition(Map<Integer, Long> currentOffsets)
+  {
+    return currentOffsets
+        .entrySet()
+        .stream()
+        .collect(
+            Collectors.toMap(
+                Map.Entry::getKey,
+                e -> latestOffsetsFromKafka != null
+                     && latestOffsetsFromKafka.get(e.getKey()) != null
+                     && e.getValue() != null
+                     ? latestOffsetsFromKafka.get(e.getKey()) - e.getValue()
+                     : null
+            )
+        );
+  }
+
+  private Runnable emitLag()
+  {
+    return () -> {
+      try {
+        Map<Integer, Long> highestCurrentOffsets = getHighestCurrentOffsets();
+
+        if (latestOffsetsFromKafka == null) {
+          throw new ISE("Latest offsets from Kafka have not been fetched");
         }
-        catch (Exception e) {
-          log.warn(e, "Unable to compute Kafka lag");
+
+        if (!latestOffsetsFromKafka.keySet().equals(highestCurrentOffsets.keySet())) {
+          log.warn(
+              "Lag metric: Kafka partitions %s do not match task partitions %s",
+              latestOffsetsFromKafka.keySet(),
+              highestCurrentOffsets.keySet()
+          );
         }
+
+        long lag = getLagPerPartition(highestCurrentOffsets)
```

Comment:
How about doing math.max with 0 when computing the total lag across all partitions, but allowing negatives when displaying individual partition lag? It makes them potentially inconsistent with each other but there is an argument for it. Total lag is more likely to be used for monitoring delay (where negatives are best treated as 0), and individual partition lag is more likely to be used for debugging finicky per partition issues (where "raw" un-massaged data is more useful).

If you do that, definitely document it, so people know what to expect.